File: src/integrator/LBInitPeriodicForce.py
#  Copyright (C) 2012-2016
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008-2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#  
#  This file is part of ESPResSo++.
#  
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>. 


r"""
    
This class allows to set or add external periodic forces 
(*lattice units*) to the LB-fluid. At first, one has to create a force object and then 
set or add this force to the system.

.. note::

    Please note, that a periodic (sin-like) force acts in *z*-direction as a function 
    of *x*. The *z*-component of the force provides therefore the amplitude of the
    sin-modulation. The *x*- and *y*-components of the specified force interpreted as
    body forces in corresponding directions.
    
Example to set external sin-like force.

>>> ampFz = 0.0001
>>> Fx = Fy = 0.
>>> extForceToSet = Real3D( Fx, Fy, ampFz )
>>> lbforceSin = espressopp.integrator.LBInitConstForce(system,lb)
>>> lbforceSin.setForce( extForceToSet )

Example to add external sin-like force. 

>>> ampFz = 0.0005
>>> Fx = Fy = 0.
>>> extForceToAdd = Real3D( Fx, Fy, ampFz )
>>> lbforceSin = espressopp.integrator.LBInitConstForce(system,lb)
>>> lbforceSin.addForce( extForceToAdd )


"""
from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.integrator.LBInit import *
from _espressopp import integrator_LBInit_PeriodicForce

class LBInitPeriodicForceLocal(LBInitLocal, integrator_LBInit_PeriodicForce):
    def __init__(self, system, latticeboltzmann):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, integrator_LBInit_PeriodicForce, system, latticeboltzmann)

if pmi.isController :
    class LBInitPeriodicForce(LBInit):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.integrator.LBInitPeriodicForceLocal',
            pmicall = [
                       "setForce", 
                       "addForce"]
            )

File: src/integrator/LBInitPeriodicForce.cpp
/*
  Copyright (C) 2012-2016
      Max Planck Institute for Polymer Research
  Copyright (C) 2008-2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "python.hpp"
#include <iomanip>
#include "LBInitPeriodicForce.hpp"

namespace espressopp {
  namespace integrator {
    LBInitPeriodicForce::LBInitPeriodicForce(shared_ptr<System> system,
                                             shared_ptr< LatticeBoltzmann > latticeboltzmann)
    : LBInit(system, latticeboltzmann) {
    }

    /* do nothing with initial populations */
    void LBInitPeriodicForce::createDenVel (real _rho0, Real3D u0) {}

    /* set external forces */
    void LBInitPeriodicForce::setForce( Real3D _force ) {
      int _printId = 0;
      int _offset = latticeboltzmann->getHaloSkin();
      Int3D _Ni = latticeboltzmann->getNi();                // system size in lattice node
      Int3D _myNi = latticeboltzmann->getMyNi();          // my local nodes
      Int3D _globIdx = latticeboltzmann->findGlobIdx();   // first lb site global index

      real ampFz = _force[2]; // amplitude of sin force in z-direction
      real perCompX;          // periodicity function with x-dir dependence
      Real3D perF;            // gravity forces in x- and y-dir; in z: amplitude
                              // of force periodic with sin in x, i.e.
                              // perForce_z = ampFz * sin ( x );

      // loop over real lattice nodes //
      real coef = 2. * M_PI / _Ni[0];
      for (int i = _offset; i < _myNi[0] - _offset; i++) {
        perCompX = sin( coef * (_globIdx[0] + i - _offset) );
        perF = Real3D (_force[0], _force[1], ampFz * perCompX);

        for (int j = _offset; j < _myNi[1] - _offset; j++) {
          for (int k = _offset; k < _myNi[2] - _offset; k++) {
            latticeboltzmann->setExtForceLoc( Int3D(i,j,k), perF );
          }
        }
      }

      // take care of the force flag for lb //
      if (_force.sqr() > ROUND_ERROR_PREC) {
        _printId = 1;
        latticeboltzmann->setDoExtForce( true );
        printForce(_force, _printId );
      }
    }

    void LBInitPeriodicForce::addForce(Real3D _force) {
      int _printId = 0;
      int _offset = latticeboltzmann->getHaloSkin();
      Int3D _Ni = latticeboltzmann->getNi();				// system size in lattice node
      Int3D _myNi = latticeboltzmann->getMyNi();		// my local nodes
      Int3D _globIdx = latticeboltzmann->findGlobIdx();// first lb site global index

      bool fOnSite = false, fTot = false;
      real ampFz = _force[2]; // amplitude of sin force in z-direction
      real perCompX;          // periodicity function with x-dir dependence
      Real3D newF, totF;      // gravity forces in x- and y-dir; in z: amplitude
                              // of force periodic with sin in x, i.e.
                              // perForce_z = ampFz * sin ( x );

      // loop over real lattice nodes //
      real coef = 2. * M_PI / _Ni[0];
      for (int i = _offset; i < _myNi[0] - _offset; i++) {
        perCompX = sin( coef * (_globIdx[0] + i - _offset) );
        newF = Real3D (_force[0], _force[1], ampFz * perCompX);

        for (int j = _offset; j < _myNi[1] - _offset; j++) {
          for (int k = _offset; k < _myNi[2] - _offset; k++) {
            totF = latticeboltzmann->getExtForceLoc( Int3D(i,j,k) ) + newF;

            // check if total force is greater than zero
            if (totF.sqr() >  ROUND_ERROR_PREC) {
              fOnSite = true;
              _printId = 2;
              latticeboltzmann->setExtForceLoc( Int3D(i,j,k), totF );
            } else {
              latticeboltzmann->setExtForceLoc( Int3D(i,j,k), Real3D(0.) );
            }

            fTot = fTot | fOnSite; // if at least one site has force -> fTot true
          }
        }
      }

      // take care of the force flag for lb //
      latticeboltzmann->setDoExtForce( fTot );
      printForce(_force, _printId );
    }

    void LBInitPeriodicForce::printForce(Real3D _force, int _printId)
    {
            if (mpiWorld->rank() == 0) {
            using namespace std;

            cout << "External force has been ";
                if (_printId == 0)
              cout << "cancelled. It is now zero.\n" ;
                else if (_printId == 1)
              cout << "set." << endl << "It is a harmonic force with amplitude: "
              << setprecision(5) << _force << endl;
                else
              cout << "added." << endl << "It is a harmonic force with amplitude: "
              << setprecision(5) << _force << endl;

            cout << "-------------------------------------" << endl;

            } else {
                // do nothing
            }
    }

    void LBInitPeriodicForce::registerPython() {
      using namespace espressopp::python;

      class_<LBInitPeriodicForce, bases< LBInit > >
          ("integrator_LBInit_PeriodicForce",	init< shared_ptr< System >,
                                                                                        shared_ptr< LatticeBoltzmann > >())
          .def("setForce", &LBInitPeriodicForce::setForce)
          .def("addForce", &LBInitPeriodicForce::addForce)
      ;
    }
  }
}
